.teamcity/_self/lib/customBuildType/E2EBuildType.kt

```diff
@@ -0,0 +1,170 @@
+import jetbrains.buildServer.configs.kotlin.v2019_2.failureConditions.failOnMetricChange
+import jetbrains.buildServer.configs.kotlin.v2019_2.triggers.schedule
+import jetbrains.buildServer.configs.kotlin.v2019_2.triggers.vcs
+
+/**
+ * This class provides a template to easily create end-to-end build configurations.
+ *
+ * To use, import then invoke the `E2EBuildType` class instance with the parameters
+ * required, as outlined in the class definition.
+ *
+ * @param buildId Human-readable value, single string.
+ * @param buildUuid UUID value generated by TeamCity and obtainable via the web UI.
+ * 		Refer to https://fieldguide.automattic.com/quality-engineering-in-calypso/running-tests/.
+ * @param buildName Human-readable display name. Primarily used in the GUI.
+ * @param buidDescription Human-readable description. Primarily used in the GUI.
+ * @param concurrentBuilds Set the number of simultaneous builds. Defaults to 0 (infinite).
+ * @param getCalypsoLiveURL String containing the commands to be run in order to obtain the live URL. Only used for Calypso E2E.
+ * @param testGroup String corresponding to an existing group defined for Jest Runner Group.
+ * @param buildParams Environment variables and other parameters to set for the build.
+ * @param buildFeatures Features to enable on top of the default feature set (perfmon, commitStatusPublisher).
+ * @param buildTriggers Rules to trigger the build. By default, no triggers are defined.
+ * @param buildDepedencies If the build configuration depends on another existing build configuration, define it here.
+ */
+open class E2EBuildType(
+	var buildId: String,
+	var buildUuid: String,
+	var buildName: String,
+	var buildDescription: String,
+	var concurrentBuilds: Int = 0,
+	var getCalypsoLiveURL: String = "",
+	var testGroup: String,
+	var buildParams: ParametrizedWithType.() -> Unit = {},
+	var buildFeatures: BuildFeatures.() -> Unit,
+	var buildTriggers: Triggers.() -> Unit = {},
+	var buildDependencies: Dependencies.() -> Unit = {},
+
+): BuildType() {
+	init {
+		val concurrentBuilds = concurrentBuilds
+		val getCalypsoLiveURL = getCalypsoLiveURL
+		val testGroup = testGroup
+		val buildParams = buildParams
+		val buildFeatures = buildFeatures
+		val buildTriggers = buildTriggers
+		val buildDependencies = buildDependencies
+		val params = params
+
+		id( buildId )
+		uuid = buildUuid
+		name = buildName
+		description = buildDescription
+		maxRunningBuilds = concurrentBuilds
+
+		artifactRules = """
+			logs.tgz => logs.tgz
+			screenshots => screenshots
+			trace => trace
+		""".trimIndent()
+
+		vcs {
+			root(Settings.WpCalypso)
+			cleanCheckout = true
+		}
+
+		params {
+			param("env.NODE_CONFIG_ENV", "test")
+			param("env.PLAYWRIGHT_BROWSERS_PATH", "0")
+			param("env.TEAMCITY_VERSION", "2021")
+			param("env.HEADLESS", "false")
+			param("env.LOCALE", "en")
+			param("env.DEBUG", "pw:api")
+			buildParams()
```

Comment:
I've chosen in this refactor to use environment variables by default for all parameters.

Though several weeks has elapsed, I recall the couple of factors that led to this:
- references to configuration parameters (unprefixed params) were failing, with TeamCity stating that it was undefined.
- may (if not all) of the parameters defined here end up being exported as an environment variable later in the script.